src/couch_file.erl

```diff
@@ -524,6 +528,214 @@ handle_info({'EXIT', _, Reason}, Fd) ->
     {stop, Reason, Fd}.
 
 
+find_last_header(Fd, FileSize) ->
+    find_header_in_chunks(Fd, start_pos(FileSize), 1).
+
+start_pos(FileSize) when FileSize rem ?SIZE_BLOCK =:= 0 ->
+    FileSize;
+start_pos(FileSize) ->
+    %% set position to nearest block boundary past eof
+    %% if file size isn't multiple of block size
+    (FileSize div ?SIZE_BLOCK) * ?SIZE_BLOCK + ?SIZE_BLOCK.
+
+chunk_max_size() ->
+    config:get_integer("couchdb", "chunk_max_size", ?DEFAULT_CHUNK_MAX_SIZE).
+
+%% chunk size exponentially increases by iteration, up to some maximum,
+%% and should never exceed the current position in the file
+chunk_size(Pos, Multiplier) ->
+    Size = min(Multiplier * ?SIZE_BLOCK, chunk_max_size()),
+    min(Pos, Size).
+
+multiplier(Size, Multiplier) ->
+    case Size < chunk_max_size() of
+        true ->
+            Multiplier * config:get_integer(
+                "couchdb", "chunk_exponent_base", ?DEFAULT_CHUNK_EXPONENT_BASE);
+        false ->
+            Multiplier
+    end.
+
+find_header_in_chunks(_Fd, Pos, _Multiplier) when Pos < 0 ->
+    no_valid_header;
+find_header_in_chunks(Fd, Pos, Multiplier) ->
+    Size = chunk_size(Pos, Multiplier),
+    case Size > 0 of
+        false ->
+            no_valid_header;
+        true ->
+            {ok, Chunk} = file:pread(Fd, Pos - Size, Size),
+            case find_header_in_chunk(Fd, Pos, Chunk, Size - ?SIZE_BLOCK) of
+                {ok, HeaderBin, _Offset} ->
+                    %% io:format("found header at ~p multiplier ~p chunksize ~p~n",
+                    %%     [Pos - Size + _Offset, Multiplier, Size]),
+                    {ok, HeaderBin};
+                not_found ->
+                    %% tell kernel we don't need that chunk any more
+                    %% to minimize disk cache pollution
```

Comment:
Did we talk about whether this actually works or not? I seem to remember a conversation about fadvise ignoring various calls. I ask because we'd also want to see if this is supported on windows (or at least doesn't throw an error).